File: pyNA/src/noise_src_py/spl.py
import numpy as np

def spl(levels, msap_prop: np.ndarray, rho_0: np.ndarray, c_0: np.ndarray) -> np.ndarray:
    """
    Compute sound pressure level.

    :param levels: pyNA component computing noise levels
    :type levels: Levels
    :param msap_prop: mean-square acoustic pressure [-]
    :type msap_prop: np.ndarray [n_t, settings.N_f]
    :param rho_0: ambient density [kg/m3]
    :type rho_0: np.ndarray [n_t]
    :param c_0: ambient speed of sound [m/s]
    :type c_0: np.ndarray [n_t]

    :return: spl [dB]
    :rtype: np.ndarray [n_t, settings.N_f]

    """
    # Load options
    n_t = levels.options['n_t']

    # Compute SPL
    spl = 10*np.log10(msap_prop) + 20.*np.log10(np.reshape(rho_0, (n_t, 1)) * np.reshape(c_0, (n_t, 1)) ** 2.)

    # Clip values to avoid (spl < 0)
    spl = spl.clip(min=1e-99)

    return spl

File: pyNA/src/noise_src_jl/spl.jl
function f_spl!(spl, x)

    # x = [c_0, rho_0]
    # y = [spl]
    
    # Compute SPL
    # Remove all SPL below 0
    spl .= 10*log10.(spl) .+ 20*log10(x[1]^2*x[2])
    
    for i in 1:1:size(spl)[1]
        if spl[i] < 1e-99
            spl[i] = 1e-99
        end
    end

end

